### PR TITLE
feature(scheduler): won't run Filter if PreFilter returned a Skip status

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

Change the framework so that it doesn't run plugins's Filter() if its PreFilter() returned a Skip status.
For example, nodeAffinity can return Skip in PreFilter if the pod doesn't specify any node selector or affinity.
This skip status is basically regarded as "Success" and you needs to distinguish Skip from Success by `IsSkip` func, not `IsSuccess` func if you need to refer to Skip status.

https://github.com/kubernetes/kubernetes/pull/112637 got reverted because of the bug. 🙏 
This PR also changes NodeAffinity PreFilter to return Skip so that we can ensure the same bug doesn't exist in this PR anymore. It's just a starting point, once this PR gets merged, I'll work on changing other PreFilter plugins to return Skip appropriately.

---

This PR is composed of three commits:
- First one: the same commit as https://github.com/kubernetes/kubernetes/pull/112637
- The second one: the patch for the bug + unit test.
- The third one: support skip in NodeAffinity.

#### Which issue(s) this PR fixes:
Part of https://github.com/kubernetes/kubernetes/issues/107556
Part of https://github.com/kubernetes/kubernetes/issues/110643

#### Does this PR introduce a user-facing change?

```release-note
Scheduler doesn't run plugin's Filter method when its PreFilter method returned a Skip status.
In other words, your PreFilter/Filter plugin can return a Skip status in PreFilter if the plugin does nothing in Filter for that Pod.
Scheduler skips NodeAffinity Filter plugin when NodeAffinity Filter plugin has nothing to do with a Pod.
It may affect some metrics values related to the NodeAffinity Filter plugin.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/assign @alculquicondor 
🙏

File: pkg/scheduler/schedule_one_test.go [hunk 1]
```diff
@@ -1979,6 +1979,42 @@ func TestSchedulerSchedulePod(t *testing.T) {
...
+				st.RegisterPreFilterPlugin(
+					"FakePreFilter1",
+					st.NewFakePreFilterPlugin("FakeFilter1", nil, nil),
+				),
+				st.RegisterFilterPlugin(
+					"FakeFilter1",
+					st.NewFakeFilterPlugin(map[string]framework.Code{
+						"node1": framework.Unschedulable,
+					}),
+				),
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one_test.go:1995]: do we need these?

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go [hunk 2]
```diff
@@ -94,8 +94,15 @@ func (pl *NodeAffinity) PreFilter(ctx context.Context, cycleState *framework.Cyc
 	affinity := pod.Spec.Affinity
 	if affinity == nil ||
 		affinity.NodeAffinity == nil ||
-		affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil ||
-		len(affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0 {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:98]: shouldn't we preserve this line here?

So that if there is no `RequiredDuringSchedulingIgnoredDuringExecution` and no `addedNodeSelector` and no `.Spec.NodeSelector`, we also return Skip.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go [hunk 3]
```diff
@@ -45,15 +45,18 @@ func TestNodeAffinity(t *testing.T) {
...
+			wantStatus:          nil, // Actually, the scheduler will skip NodeAffinity Filter as PreFilter returns Skip.
 		},
 		{
 			name: "missing labels",
 			pod: st.MakePod().NodeSelector(map[string]string{
 				"foo": "bar",
 			}).Obj(),
-			wantStatus: framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPod),
+			wantStatus:       framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPod),
+			disablePreFilter: true,
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:59]: why? Same question for the rest of the cases.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go [hunk 4]
```diff
@@ -45,15 +45,18 @@ func TestNodeAffinity(t *testing.T) {
 		disablePreFilter    bool
 	}{
 		{
-			name: "no selector",
-			pod:  &v1.Pod{},
+			name:                "no selector and affinity",
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:48]: how is this different from test case "Pod with no Affinity will schedule onto a node" below?

File: pkg/scheduler/framework/interface.go [hunk 5]
```diff
@@ -203,9 +203,9 @@ func (s *Status) AppendReason(reason string) {
 	s.reasons = append(s.reasons, reason)
 }
 
-// IsSuccess returns true if and only if "Status" is nil or Code is "Success".
+// IsSuccess returns true if and only if "Status" is nil or Code is "Success" or "Skip".
 func (s *Status) IsSuccess() bool {
-	return s.Code() == Success
+	return s.Code() == Success || s.Code() == Skip
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/interface.go:208]: This method is generally used like `if !s.IsSuccess()`, in which it makes sense. Maybe we should change the name? But I think `Skip` is not too different from Success.

File: pkg/scheduler/schedule_one_test.go (hunk 1 above)
Comment (sanposhiho) [reply pkg/scheduler/schedule_one_test.go:1995]: Here, I want to make sure that all other plugins are executed even if some plugins return skip.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go (hunk 3 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:59]: Removed all `disablePreFilter`.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:98]: You mean we can add `len(affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0` here, right? I think that's true. Fixed

File: pkg/scheduler/framework/interface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go:208]: I feel the same as @alculquicondor. Skip is also kinda happy path.
I'm not against renaming, but is there any good name for the func? No good idea from my poor English vocab. 😓

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/interface.go:208]: We not only depend on `!s.IsSuccess()`, also `s.IsSuccess()`, see here (also other places)https://github.com/kubernetes/kubernetes/blob/17bf864c1fc20e4badc9a6e0659c5190310462e1/pkg/scheduler/schedule_one.go#L514-L521

This is not for PreFilter, but we now changed the underlying  meanings of `Success`. If the code is Skip, I don't think we should process with this. Maybe we should add another method like `IsPassed()` means we can continue with the logic that follows, both success or skip.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go:208]: Keeping `IsSuccess` and creating another func like `IsPassed`(= `Success` and `Skip`) makes sense to me.

In `pkg/scheduler/schedule_one.go`, we currently need to use `IsSkip` to distinguish `Skip` from `IsSuccess`(= `Success` and `Skip`), and a new `IsSuccess` (= only `Success`) would be useful there.
We can just change all other places where using `IsSuccess` to use `IsPassed`

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/interface.go:208]: sgtm

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go:213]: 320a11415eaf2a9c6b7d46255c8bec06d156c951.

File: pkg/scheduler/framework/runtime/framework.go [hunk 6]
```diff
@@ -617,6 +620,9 @@ func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framewor
 			}
 			return nil, framework.AsStatus(fmt.Errorf("running PreFilter plugin %q: %w", pl.Name(), s.AsError())).WithFailedPlugin(pl.Name())
 		}
+		if s.IsSkip() {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:623]: I may not correctly grasp what you mean, but if a returned status isn't a success, then it doesn't reach here.

**Comment by alculquicondor** [comment]
@sanposhiho can you prioritize this? There are multiple contributors waiting on this.

**Comment by sanposhiho** [comment]
@alculquicondor I think It's already ready to get reviews.

**Comment by alculquicondor** [comment]
needs rebase though

**Comment by sanposhiho** [comment]
rebased

**Comment by kerthcet** [comment]
UT failed due to https://github.com/kubernetes/kubernetes/issues/107414.

File: pkg/scheduler/framework/interface.go [hunk 7]
```diff
@@ -244,7 +250,7 @@ func (s *Status) AsError() error {
 // cmp.Equal.
 func (s *Status) Equal(x *Status) bool {
 	if s == nil || x == nil {
-		return s.IsSuccess() && x.IsSuccess()
+		return s.IsPassed() && x.IsPassed()
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/interface.go:253]: Can we reduce the scope only impact the preFilter? This is more safe.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go]: > Can we reduce the scope only impact the preFilter?

Fair point. reverted the commit of all replaces for the current `IsSuccess` ([5f71d63](https://github.com/kubernetes/kubernetes/pull/114125/commits/5f71d635fa4ca3e20796ca9655e12dcbb56bff1b)) and changed some `IsSuccess` which only related to PreFilter. ([c2a9b0a](https://github.com/kubernetes/kubernetes/pull/114125/commits/c2a9b0a2fab1cf26076e5577dafd2fdf51001612))

File: pkg/scheduler/framework/cycle_state.go [hunk 8]
```diff
@@ -48,6 +50,8 @@ type CycleState struct {
 	storage sync.Map
 	// if recordPluginMetrics is true, PluginExecutionDuration will be recorded for this cycle.
 	recordPluginMetrics bool
+	// SkipFilterPlugins are plugins that will be skipped in the Filter extension point.
+	SkipFilterPlugins sets.String
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/cycle_state.go]: Ah, nice catch.

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/cycle_state.go]: 9f9260e

File: pkg/scheduler/framework/interface.go [hunk 9]
```diff
@@ -231,7 +237,7 @@ func (s *Status) IsUnschedulable() bool {
 // AsError returns nil if the status is a success, a wait or a skip; otherwise returns an "error" object
 // with a concatenated message on reasons of the Status.
 func (s *Status) AsError() error {
-	if s.IsSuccess() || s.IsWait() || s.IsSkip() {
+	if s.IsPassed() || s.IsWait() || s.IsSkip() {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go]: [5f71d63](https://github.com/kubernetes/kubernetes/pull/114125/commits/5f71d635fa4ca3e20796ca9655e12dcbb56bff1b)

**Comment by alculquicondor** [comment]
Multiple scheduler tests still failing

**Comment by sanposhiho** [comment]
@alculquicondor Could you please review again?

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 10]
```diff
@@ -1412,40 +1412,152 @@ func TestPreFilterPlugins(t *testing.T) {
...
-	_, status := f.RunPreFilterPlugins(ctx, nil, nil)
-	wantStatus := framework.AsStatus(fmt.Errorf("running PreFilter plugin %q: %w", preFilter.Name(), errInjectedStatus)).WithFailedPlugin(preFilter.Name())
-	if !reflect.DeepEqual(status, wantStatus) {
-		t.Errorf("wrong status. got: %v, want:%v", status, wantStatus)
+			state := framework.NewCycleState()
+			result, status := f.RunPreFilterPlugins(ctx, state, nil)
+			if d := cmp.Diff(result, tt.wantPreFilterResult); d != "" {
+				t.Errorf("wrong status. got: %v, want: %v, diff: %s", result, tt.wantPreFilterResult, d)
+			}
+			if d := cmp.Diff(status, tt.wantStatus, cmp.Comparer(func(a, b *framework.Status) bool {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework_test.go:1537]: you shouldn't need this custom comparer after #114456

File: pkg/scheduler/schedule_one_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/schedule_one_test.go:1996]: I'm fine leaving that tests just to the runtime package, but up to you.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:98]: Ohh, that's true. Let's keep the behavior.

Btw, I checked the validation logic, and Pods with a nil []NodeSelectorTerm should be rejected actually. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4012-L4014

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go [hunk 11]
```diff
@@ -237,7 +239,8 @@ func TestNodeAffinity(t *testing.T) {
 			labels: map[string]string{
 				"foo": "bar",
 			},
-			wantStatus: framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPod),
+			wantPreFilterStatus: framework.NewStatus(framework.Skip),
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:242]: I make it return nil in this case, but actually it's rejected by validation webhook. let's move to https://github.com/kubernetes/kubernetes/pull/114125/files#r1056150739.

File: pkg/scheduler/schedule_one_test.go (hunk 1 above)
Comment (sanposhiho) [reply pkg/scheduler/schedule_one_test.go:1996]: I'd like to keep it because if this test have existed in the past PR (reverted one), actually we wouldn't have encountered the bug. 
The runtime package's test can only confirm the behavior of specific extension point (Only either Filter or PreFilter) and it cannot confirm what if the scheduler actually run from PreFilter to Filter.

**Comment by sanposhiho** [comment]
@alculquicondor Thanks. Fixed and rebased on the latest master. PTAL after your vacation. 🌴

**Comment by sanposhiho** [comment]
/retest 
https://github.com/kubernetes/kubernetes/issues/110697

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 12]
```diff
@@ -1412,40 +1412,149 @@ func TestPreFilterPlugins(t *testing.T) {
...
-	_, status := f.RunPreFilterPlugins(ctx, nil, nil)
-	wantStatus := framework.AsStatus(fmt.Errorf("running PreFilter plugin %q: %w", preFilter.Name(), errInjectedStatus)).WithFailedPlugin(preFilter.Name())
-	if !reflect.DeepEqual(status, wantStatus) {
-		t.Errorf("wrong status. got: %v, want:%v", status, wantStatus)
+			state := framework.NewCycleState()
+			result, status := f.RunPreFilterPlugins(ctx, state, nil)
+			if d := cmp.Diff(result, tt.wantPreFilterResult); d != "" {
+				t.Errorf("wrong status. got: %v, want: %v, diff: %s", result, tt.wantPreFilterResult, d)
+			}
+			if d := cmp.Diff(status.Code(), tt.wantStatus.Code()); d != "" {
```
Comment (kerthcet) [inline pkg/scheduler/framework/runtime/framework_test.go:1537]: I think @alculquicondor suggestion is using `cmp.Diff() ` or `cmp.Equal()`, but the problem here is error is not equal. It's tricky.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go [hunk 13]
```diff
@@ -89,16 +89,24 @@ func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEvent {
...
+		affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		if pl.addedNodeSelector == nil && pod.Spec.NodeSelector == nil {
+			// NodeAffinity Filter has nothing to do with the Pod.
+			return nil, framework.NewStatus(framework.Skip)
+		}
 		return nil, nil
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:100]: >  only if pod.Spec.NodeSelector != nil, we may also want to calculate GetRequiredNodeAffinity

that's true, nice catch. 🙏

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go [hunk 14]
```diff
@@ -44,10 +44,6 @@ func TestNodeAffinity(t *testing.T) {
 		args                config.NodeAffinityArgs
 		disablePreFilter    bool
 	}{
-		{
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:47]: Because it's the almost same as `"Pod with no Affinity will schedule onto a node"`.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:98]: ahh, good find. It should be fine to remove the test case. Up to you.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 12 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework_test.go:1537]: Right, it's not equal, but does the message matter?

I don't think it matters, so we can just have `tt.wantStatusCode` instead of `tt.wantStatus`. Otherwise... we should be fixing `TestPlugin` to return a predictable error. But that can be a follow up.

File: pkg/scheduler/schedule_one.go [hunk 15]
```diff
@@ -397,7 +397,7 @@ func (sched *Scheduler) findNodesThatFitPod(ctx context.Context, fwk framework.F
 	}
 	// Run "prefilter" plugins.
 	preRes, s := fwk.RunPreFilterPlugins(ctx, state, pod)
-	if !s.IsSuccess() {
+	if !s.Passed() {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:400]: Let's imagine an scenario where all the plugins are skipped. Wouldn't the framework runtime return Success?

So this change shouldn't matter, right?

**Review by alculquicondor — COMMENTED** [review]
/approve

just a nit about the test.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 12 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework_test.go:1537]: > we can just have tt.wantStatusCode

👍

File: pkg/scheduler/schedule_one.go (hunk 15 above)
Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:400]: I have one thing need to apologize to you related to this; I realized I had misunderstood and the bug (which was the reason I reverted the previous PR) didn’t actually exist. so I should not have reverted the previous PR. I’m so sorry about this and to take up your extra time to review this PR. 🙇 🙇 

So, you are correct at this point, actually, this change doesn’t matter. Also, we probably don’t need to create `Passed()` function because any upstream function should not receive Skip.
Let me revert those useless changes.

Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go]: Can you add a test case to ensure that RunPreFilterPlugins to return Success when all plugins return Skip?

Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go]: No need to apologize. Better be safe than sorry :)

Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go]: > Can you add a test case to ensure that RunPreFilterPlugins to return Success when all plugins return Skip?

We have already had `"all PreFilter plugins returned skip",` test case in pkg/scheduler/framework/runtime/framework_test.go#L1475.

File: pkg/scheduler/framework/runtime/framework.go [hunk 16]
```diff
@@ -599,24 +599,30 @@ func (f *frameworkImpl) QueueSortFunc() framework.LessFunc {
...
 	defer func() {
 		metrics.FrameworkExtensionPointDuration.WithLabelValues(preFilter, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
 	}()
 	var result *framework.PreFilterResult
 	var pluginsWithNodes []string
+	skipPlugins := sets.New[string]()
 	for _, pl := range f.preFilterPlugins {
 		r, s := f.runPreFilterPlugin(ctx, pl, state, pod)
-		if !s.IsSuccess() {
+		if !s.IsSuccess() && !s.IsSkip() {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:616]:
```golang
if s.IsSkip() {
  skipPlugins.Insert(pl.Name())
} else if !s.IsSuccess() {
  ...
}

**Review by alculquicondor — COMMENTED** [review]
/approve

File: pkg/scheduler/framework/runtime/framework.go (hunk 16 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework.go]: Actually, wouldn't it be better like this?

```golang
if s.IsSkip() {
  skipPlugins.Insert(pl.Name())
  continue
}
```

A Skip code should not have any relevant information in `r`, right?

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go]: done

Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework.go]: Oops, you were too fast. I had another thought above.

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go]: > A Skip code should not have any relevant information in r, right?

Right. Let me fix

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go]: Done.

File: pkg/scheduler/framework/interface.go [hunk 17]
```diff
@@ -385,6 +385,8 @@ type PreFilterPlugin interface {
 	// plugins must return success or the pod will be rejected. PreFilter could optionally
 	// return a PreFilterResult to influence which nodes to evaluate downstream. This is useful
 	// for cases where it is possible to determine the subset of nodes to process in O(1) time.
+	// When it returns Skip status, returned PreFilterResult and other fields in status are just ignored,
+	// and coupled Filter plugin will be skipped in this scheduling cycle.
 	PreFilter(ctx context.Context, state *CycleState, p *v1.Pod) (*PreFilterResult, *Status)
 	// PreFilterExtensions returns a PreFilterExtensions interface if the plugin implements one,
```
Comment (alculquicondor) [inline pkg/scheduler/framework/interface.go:391]: Actually, another thought:

Should we explicitly skip calling PreFilterExtensions when PreFilter returned Skip?

Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go:391]: > Should we explicitly skip calling PreFilterExtensions when PreFilter returned Skip?

Good catch.. Yes, I believe we should skip. 

- we only use PreFilterExtensions in `SelectVictimsOnNode ` func in the default_preemption.
  - https://github.com/kubernetes/kubernetes/blob/2d534e4bea7c7ef412a03c921867d9027f6e489f/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L144-L161
- The `SelectVictimsOnNode` run Filter plugins via `RunFilterPluginsWithNominatedPods` func.
  - https://github.com/kubernetes/kubernetes/blob/2d534e4bea7c7ef412a03c921867d9027f6e489f/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L186
- The `RunFilterPluginsWithNominatedPods` func calls `RunFilterPlugins` to run Filter plugins.
  - https://github.com/kubernetes/kubernetes/blob/2d534e4bea7c7ef412a03c921867d9027f6e489f/pkg/scheduler/framework/runtime/framework.go#L836
- In `RunFilterPlugins`, Filter plugins in `SkipFilterPlugins` of CycleState are skipped.


So, it doesn't make sense to call `PreFilterExtensions.AddPod/RemovePod` of `SkipFilterPlugins` in the default_preemption because those plugins aren't executed after that.
Let me fix.

Comment (sanposhiho) [reply pkg/scheduler/framework/interface.go:391]: > we only use PreFilterExtensions in SelectVictimsOnNode  func in the default_preemption.

It's wrong, we also use `PreFilterExtensions().AddPod()` in `addNominatedPods` as well, but the conclusion is the same since `addNominatedPods` is called only in `RunFilterPluginsWithNominatedPods`.
https://github.com/kubernetes/kubernetes/blob/2d534e4bea7c7ef412a03c921867d9027f6e489f/pkg/scheduler/framework/runtime/framework.go#L863

**Comment by alculquicondor** [comment]
/approve

Please squash

**Comment by sanposhiho** [comment]
Squashed. Thanks @alculquicondor @kerthcet for the long-time effort in reviewing this.

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 18]
```diff
@@ -1402,9 +1402,11 @@ func TestPreFilterPlugins(t *testing.T) {
...
 			t.Fatalf("Failed to create framework for testing: %v", err)
 		}
-		f.RunPreFilterPlugins(ctx, nil, nil)
-		f.RunPreFilterExtensionAddPod(ctx, nil, nil, nil, nil)
-		f.RunPreFilterExtensionRemovePod(ctx, nil, nil, nil, nil)
+		state := framework.NewCycleState()
+
+		f.RunPreFilterPlugins(ctx, state, nil)
+		f.RunPreFilterExtensionAddPod(ctx, state, nil, nil, nil)
+		f.RunPreFilterExtensionRemovePod(ctx, state, nil, nil, nil)
```
Comment (sanposhiho) [inline pkg/scheduler/framework/runtime/framework_test.go:1409]: Just added these changes to prevent panic in `RunPreFilterExtensionAddPod` and `RunPreFilterExtensionRemovePod`. 🙏

**Review by alculquicondor — COMMENTED** [review]
/lgtm
Thanks!

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 19]
```diff
@@ -1421,39 +1423,310 @@ func TestPreFilterPlugins(t *testing.T) {
...
-	preFilter := &TestPlugin{
-		name: preFilterPluginName,
-		inj:  injectedResult{PreFilterStatus: int(framework.Error)},
+func TestRunPreFilterPlugins(t *testing.T) {
+	tests := []struct {
+		name                string
+		plugins             []*TestPlugin
+		wantPreFilterResult *framework.PreFilterResult
+		wantSkippedPlugins  sets.Set[string]
+		wantStatusCode      framework.Code
```
Comment (pacoxu) [inline pkg/scheduler/framework/runtime/framework_test.go:1432]: /hold
for go format CI failure

**Comment by sanposhiho** [comment]
Oops, sorry for that. Formated. 🙏

**Comment by sanposhiho** [comment]
/retest

https://github.com/kubernetes/kubernetes/issues/110697

**Comment by sanposhiho** [comment]
/retest

https://github.com/kubernetes/kubernetes/issues/114438
